### PR TITLE
Fixed and clarified comments and parameter names in index.go

### LINKED ISSUE #69319: k8s.io/client-go/tools/cache/index.go confuses keys with values
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:
I wrote some client code based on the usage of the word "key" in index.go, and my code was buggy.  Then I looked more closely at index.go, and realized that it has a dichotomy but botches the naming and discussion of it.  The dichotomy is between what that file sometimes calls "values" and some of what it calls "keys".  An `IndexFunc` maps an object to a set of "values".  The `IndexKeys` method of an `Indexer` maps `(a, b)` to a slice of keys and an error, where `a` names an index and `b` is a "value".  Specifically, the resulting slice of keys are those whose corresponding object is mapped by the named `IndexFunc` to a set of "values" that includes `b`.  But the name used for this second parameter (which I have called `b` here) is `indexKey`.  Instead it should be something more like `indexValue`.  The `ByIndex` method has the same parameter naming problem.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): all of them
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design

/kind documentation

> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Clarified the comments and parameter names in the Indexer interface to
consistently use a dichotomy between "storage keys" and "indexed
values".
Also replaced vague wording about "match" with a precise characterization (intersection of set of indexed values).
Also, fixed the comment on IndexFunc to say that it returns multiple indexed values.

**Which issue(s) this PR fixes**:

Fixes #69319 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-architecture-bugs

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-bugs

**Review by logicalhan — APPROVED** [review]
/lgtm 

Thanks for picking this back up!

**Comment by MikeSpreitzer** [comment]
@mikedanese , @eparis : ping?

**Comment by MikeSpreitzer** [comment]
Fixed oversight in inventory of sorts of strings.

File: staging/src/k8s.io/client-go/tools/cache/index.go [hunk 1]
```diff
@@ -42,7 +53,7 @@ type Indexer interface {
 	AddIndexers(newIndexers Indexers) error
 }
 
-// IndexFunc knows how to provide an indexed value for an object.
+// IndexFunc knows how to extract the set of indexed values for an object.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/index.go]: `IndexFunc knows how to extract the set of index keys for an object.`

File: staging/src/k8s.io/client-go/tools/cache/index.go [hunk 2]
```diff
@@ -23,17 +23,28 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// Indexer is a storage interface that lets you list objects using multiple indexing functions
+// Indexer is a storage interface that lets you list objects using multiple indexing functions.
+// There are three sorts of strings here.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/client-go/tools/cache/index.go]: Oh, I see how that could be confusing.  I am a little shy of "type", as it is a technical term in golang.  How about "kind"?

File: staging/src/k8s.io/client-go/tools/cache/index.go [hunk 3]
```diff
@@ -23,17 +23,28 @@ import (
...
+// One is a storage key, as defined in the Store interface.
+// Another sort is a name of an IndexFunc.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/client-go/tools/cache/index.go]: OK

File: staging/src/k8s.io/client-go/tools/cache/index.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/cache/index.go]: sure

File: staging/src/k8s.io/client-go/tools/cache/index.go [hunk 4]
```diff
@@ -23,17 +23,28 @@ import (
...
+// The third sort of string is an "indexed value", produced by an IndexFunc.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/client-go/tools/cache/index.go]: I am really surprised by this objection.  I never got the implication that an "indexed value" had to be the value of a field of the object.  Note also that the original language defining an "IndexFunc" describes its outputs as "indexed values" (modulo the bug of using singular where plural is meant).  How about keeping "indexed value" and adding a clarification like the following:
```
// The third kind of string is an "indexed value", which is produced by an IndexFunc and can be a field value or any other string computed from the object
```

**Comment by MikeSpreitzer** [comment]
@liggitt : updated, PTAL

**Pull request merged**